pkg/controllers/metrics/node/controller.go

```diff
@@ -0,0 +1,338 @@
+	}
+	newlabels, err := c.generateLabels(node)
+	if err != nil {
+		logging.FromContext(ctx).Errorf("Failed to generate labels: %s", err.Error())
+		return reconcile.Result{}, err
+	}
+
+	if err := c.updateGaguges(ctx, node, newlabels); err != nil {
+		logging.FromContext(ctx).Errorf("Failed to update gauges: %s", err.Error())
+		return reconcile.Result{}, err
+	}
+	c.LabelsMap[req.NamespacedName] = newlabels
+
+	return reconcile.Result{}, nil
+}
+
+func (c *Controller) Register(ctx context.Context, m manager.Manager) error {
+	err := controllerruntime.
+		NewControllerManagedBy(m).
+		Named("nodemetrics").
+		For(&v1.Node{}).
+		Watches(
+			// Reconcile all nodes related to a provisioner when it changes.
+			&source.Kind{Type: &v1alpha5.Provisioner{}},
+			handler.EnqueueRequestsFromMapFunc(func(o client.Object) (requests []reconcile.Request) {
+				nodes := &v1.NodeList{}
+				if err := c.KubeClient.List(ctx, nodes, client.MatchingLabels(map[string]string{v1alpha5.ProvisionerNameLabelKey: o.GetName()})); err != nil {
+					logging.FromContext(ctx).Errorf("Failed to list nodes when mapping expiration watch events, %s", err.Error())
+					return requests
+				}
+				for _, node := range nodes.Items {
+					requests = append(requests, reconcile.Request{NamespacedName: types.NamespacedName{Name: node.Name}})
+				}
+				return requests
+			}),
+		).
+		Watches(
+			// Reconcile node when a pod assigned to it changes.
+			&source.Kind{Type: &v1.Pod{}},
+			handler.EnqueueRequestsFromMapFunc(func(o client.Object) (requests []reconcile.Request) {
+				if name := o.(*v1.Pod).Spec.NodeName; name != "" {
+					requests = append(requests, reconcile.Request{NamespacedName: types.NamespacedName{Name: name}})
+				}
+				return requests
+			}),
+		).
+		WithOptions(controller.Options{MaxConcurrentReconciles: 10000}).
+		Complete(c)
+	return err
+}
+
+func (c *Controller) deleteGaguges(labels *prometheus.Labels) {
+	for _, gaugeVec := range c.GaugeVecMap {
+		gaugeVec.Delete(*labels)
+	}
+}
+
+// generateLabels creates the labels using the current state of the pod
+func (c *Controller) generateLabels(node *v1.Node) (*prometheus.Labels, error) {
+	metricLabels := prometheus.Labels{}
+
+	metricLabels[nodeName] = node.GetName()
+	metricLabels[nodeProvisioner] = node.Labels[v1alpha5.ProvisionerNameLabelKey]
+	metricLabels[nodeZone] = node.Labels[v1.LabelTopologyZone]
+	metricLabels[nodeArchitecture] = node.Labels[v1.LabelArchStable]
+	metricLabels[nodeCapacityType] = node.Labels[v1alpha5.LabelCapacityType]
+	metricLabels[nodeInstanceType] = node.Labels[v1.LabelInstanceTypeStable]
+	metricLabels[nodePhase] = string(node.Status.Phase)
+	// Add node labels
+	labels, err := json.Marshal(node.GetLabels())
+	if err != nil {
+		return nil, fmt.Errorf("marshal pod labels: %w", err)
+	}
+	metricLabels[nodeLabels] = string(labels)
+	return &metricLabels, nil
+}
+
+func (c *Controller) updateGaguges(ctx context.Context, node *v1.Node, labels *prometheus.Labels) error {
```

Comment:
It's a bit odd to me that you've structured the code differently for these two controllers, yet they do almost exactly the same thing over different resources.

Reply:
One of the major differences is that pod controller has only one gaugeVec (the pod state), but node controller needs to generates one gaugeVec for each resource type. Some resource type is pretty standard like CPU, memory but there are cloud specific type like attachable-volumes-aws-ebs. I don't want to hardcode the resource types so we dynamically generate gaugeVec on the flight instead.

Another major difference is that the pod controller doesn't really calculate a metric value (it always return 1 to indicate that there is a pod in that state). But the node controller needs to find out resource levels from nodes.